FirebaseFunctions/Tests/Unit/FunctionsErrorTests.swift

```diff
@@ -0,0 +1,172 @@
+    // The status code in the error body takes precedence over the HTTP status code.
+    let responseData =
+      #"{ "error": { "status": "OUT_OF_RANGE", "message": "TEST_ErrorMessage", "details": 123 } }"#
+        .data(using: .utf8)!
+
+    let error = FunctionsError(
+      httpStatusCode: 499,
+      body: responseData,
+      serializer: FunctionsSerializer()
+    )
+
+    guard let error else { return XCTFail("Unexpected `nil` value") }
+
+    let nsError = error as NSError
+    XCTAssertEqual(nsError.domain, "com.firebase.functions")
+    XCTAssertEqual(nsError.code, 11)
+    XCTAssertEqual(nsError.localizedDescription, "TEST_ErrorMessage")
+    XCTAssertEqual(nsError.userInfo.count, 2)
+    XCTAssertEqual(nsError.userInfo["details"] as? Int, 123)
+  }
+
+  func testInitWithErrorStatusCodeAndOKErrorBody() {
+    // When the status code in the error body is `OK`, error should be `nil` regardless of the HTTP
+    // status code.
+    let responseData =
+      #"{ "error": { "status": "OK", "message": "TEST_ErrorMessage", "details": 123 } }"#
+        .data(using: .utf8)!
+
+    let error = FunctionsError(
+      httpStatusCode: 401,
+      body: responseData,
+      serializer: FunctionsSerializer()
+    )
+
+    XCTAssertNil(error)
+  }
+
+  func testInitWithErrorStatusCodeAndIncompleteErrorBody() {
+    // The error name is not in the body; it should be inferred from the HTTP status code.
+    let responseData = #"{ "error": { "message": "TEST_ErrorMessage", "details": null } }"#
+      .data(using: .utf8)!
+
+    let error = FunctionsError(
+      httpStatusCode: 403,
+      body: responseData,
+      serializer: FunctionsSerializer()
+    )
+
+    guard let error else { return XCTFail("Unexpected `nil` value") }
+
+    let nsError = error as NSError
+    XCTAssertEqual(nsError.domain, "com.firebase.functions")
+    XCTAssertEqual(nsError.code, 7) // `permissionDenied`, inferred from the HTTP status code
+    XCTAssertEqual(nsError.localizedDescription, "TEST_ErrorMessage")
+    XCTAssertEqual(nsError.userInfo.count, 2)
+    XCTAssertEqual(nsError.userInfo["details"] as? NSNull, NSNull())
+  }
+
+  func testInitWithErrorStatusCodeAndInvalidErrorBody() {
+    // An unsupported status code in the error body should result in the rest of the body ignored.
+    let responseData =
+      #"{ "error": { "status": "TEST_UNKNOWN_ERROR", "message": "TEST_ErrorMessage", "details": 123 } }"#
+        .data(using: .utf8)!
+
+    let error = FunctionsError(
+      httpStatusCode: 503,
+      body: responseData,
+      serializer: FunctionsSerializer()
+    )
+
+    guard let error else { return XCTFail("Unexpected `nil` value") }
+
+    let nsError = error as NSError
+    XCTAssertEqual(nsError.domain, "com.firebase.functions")
+    // Currently, `internal` is used as the fallback error code. Is this correct?
+    // Seems like we could get more information from the HTTP status code in such cases.
```

Comment:
Here, this behavior results in data loss. And in general, `internal` seems somewhat confusing for unknown error codes (especially when there’s a separate `unknown` error case). I guess this should be looked into someday.

Reply:
Agreed. This is fine as is, but maybe the `unknown` case should have an associated value to provide a status code with more info. I'm not sure how if it would affect the ObjC interface. In practice, this may not be a rare/non issue.